nengo/transforms.py

```diff
@@ -0,0 +1,290 @@
+
+    @property
+    def size_in(self):
+        """Expected size of input to transform"""
+        raise NotImplementedError()
+
+    @property
+    def size_out(self):
+        """Expected size of output from transform"""
+        raise NotImplementedError()
+
+
+class ChannelShapeParam(ShapeParam):
+    """A parameter where the value must be a shape with channels."""
+
+    def coerce(self, transform, shape):
+        if isinstance(shape, ChannelShape):
+            if shape.channels_last != transform.channels_last:
+                raise ValidationError(
+                    "transform has channels_last=%s, but input shape has "
+                    "channels_last=%s" % (transform.channels_last,
+                                          shape.channels_last),
+                    attr=self.name, obj=transform)
+            super(ChannelShapeParam, self).coerce(transform, shape.shape)
+        else:
+            super(ChannelShapeParam, self).coerce(transform, shape)
+            shape = ChannelShape(shape, channels_last=transform.channels_last)
+        return shape
+
+
+class Dense(Transform):
+    """
+    A dense transformation between an input and output signal.
+
+    Parameters
+    ----------
+    shape : tuple of int
+        The shape of the dense matrix: ``(size_out, size_in)``.
+    init : :class:`.Distribution` or array_like, optional
+        A Distribution used to initialize the transform matrix, or a concrete
+        instantiation for the matrix.  If the matrix is square we also allow a
+        scalar (equivalent to ``np.eye(n) * init``) or a vector (equivalent to
+        ``np.diag(init)``) to represent the matrix more compactly.
+    """
+
+    shape = ShapeParam("shape", length=2, low=1)
+    init = DistOrArrayParam("init")
+
+    def __init__(self, shape, init=1.0):
+        super(Dense, self).__init__()
+
+        self.shape = shape
+
+        if is_array_like(init):
+            init = np.asarray(init, dtype=np.float64)
+
+            # check that the shape of init is compatible with the given shape
+            # for this transform
+            expected_shape = None
+            if shape[0] != shape[1]:
+                # init must be 2D if transform is not square
+                expected_shape = shape
+            elif init.ndim == 1:
+                expected_shape = (shape[0],)
+            elif init.ndim >= 2:
+                expected_shape = shape
+
+            if expected_shape is not None and init.shape != expected_shape:
+                raise ValidationError(
+                    "Shape of initial value %s does not match expected "
+                    "shape %s" % (init.shape, expected_shape), "init")
+
+        self.init = init
+
+    def sample(self, rng=np.random):
+        if isinstance(self.init, Distribution):
+            return self.init.sample(*self.shape, rng=rng)
+
+        return self.init
+
+    @property
+    def init_shape(self):
+        """The shape of the initial value."""
```

Comment:
It would be different due to the broadcasting we allow (e.g. passing `transform=1.0` instead of the full weight matrix).  So in that case `init_shape` would be `()`, but `shape` would be `(size_out, size_in)`.